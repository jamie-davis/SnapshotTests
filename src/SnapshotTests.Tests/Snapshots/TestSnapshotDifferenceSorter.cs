using System.Configuration;
using System.Linq;
using NUnit.Framework;
using SnapshotTests.Snapshots;
using SnapshotTests.Tests.Snapshots.TestSupportUtils;
using TestConsoleLib;
using TestConsoleLib.Testing;

namespace SnapshotTests.Tests.Snapshots
{
    [TestFixture]
    public class TestSnapshotDifferenceSorter
    {
        [Test]
        public void DifferenceDataIsSorted()
        {
            //Arrange
            var collection = new SnapshotCollection();
            collection.DefineTable("TheTable").PrimaryKey("Id").Sort("Name");

            void AddRow(SnapshotBuilder snapshot, int id, string name)
            {
                var add = snapshot.AddNewRow("TheTable");
                add["Id"] = id;
                add["Name"] = name;
            }

            var beforeBuilder = collection.NewSnapshot("Before");
            AddRow(beforeBuilder, 1, "Zed");
            AddRow(beforeBuilder, 2, "Xavier");
            AddRow(beforeBuilder, 3, "Alice");

            var afterBuilder = collection.NewSnapshot("After");
            AddRow(afterBuilder, 1, "Zed+");
            AddRow(afterBuilder, 2, "Xavier+");
            AddRow(afterBuilder, 3, "Alice+");

            var before = collection.GetSnapshot("Before");
            var after = collection.GetSnapshot("After");
            var diffs = SnapshotDifferenceAnalyser.ExtractDifferences(collection, before, after);

            //Act
            var result = SnapshotDifferenceSorter.SortDifferences(collection, diffs.TableDifferences.ToList());

            //Assert
            var output = new Output();
            result.Report(output);
            output.Report.Verify();

        }

        [Test]
        public void DifferenceDataIsSortedDescending()
        {
            //Arrange
            var collection = new SnapshotCollection();
            collection.DefineTable("TheTable").PrimaryKey("Id").SortDescending("Name");

            void AddRow(SnapshotBuilder snapshot, int id, string name)
            {
                var add = snapshot.AddNewRow("TheTable");
                add["Id"] = id;
                add["Name"] = name;
            }

            var beforeBuilder = collection.NewSnapshot("Before");
            AddRow(beforeBuilder, 1, "Alice");
            AddRow(beforeBuilder, 2, "Xavier");
            AddRow(beforeBuilder, 3, "Zed");

            var afterBuilder = collection.NewSnapshot("After");
            AddRow(afterBuilder, 1, "Alice+");
            AddRow(afterBuilder, 2, "Xavier+");
            AddRow(afterBuilder, 3, "Zed+");

            var before = collection.GetSnapshot("Before");
            var after = collection.GetSnapshot("After");
            var diffs = SnapshotDifferenceAnalyser.ExtractDifferences(collection, before, after);

            //Act
            var result = SnapshotDifferenceSorter.SortDifferences(collection, diffs.TableDifferences.ToList());

            //Assert
            var output = new Output();
            result.Report(output);
            output.Report.Verify();
        }

        [Test]
        public void DifferenceDataIsSortedByMultipleColumns()
        {
            //Arrange
            var collection = new SnapshotCollection();
            collection.DefineTable("TheTable")
                .PrimaryKey("Id")
                .Sort("ColAsc1")
                .SortDescending("ColDesc1")
                .Sort("ColAsc2")
                .SortDescending("ColDesc2");

            void AddRow(SnapshotBuilder snapshot, int id, string asc1, string desc1, string asc2, string desc2)
            {
                var add = snapshot.AddNewRow("TheTable");
                add["Id"] = id;
                add["ColAsc1"] = asc1;
                add["ColDesc1"] = desc1;
                add["ColAsc2"] = asc2;
                add["ColDesc2"] = desc2;
            }

            void AddData(SnapshotBuilder snapshot, string ascData, string descData)
            {
                var id = 1;
                foreach (var colAsc1 in ascData.Split(','))
                {
                    foreach (var colDesc1 in descData.Split(','))
                    {
                        foreach (var colAsc2 in ascData.Split(','))
                        {
                            foreach (var colDesc2 in descData.Split(','))
                            {
                                AddRow(snapshot, id++, colAsc1, colDesc1, colAsc2, colDesc2);
                            }
                        }
                    }
                }
            }

            var beforeBuilder = collection.NewSnapshot("Before");
            AddData(beforeBuilder, "A,B,C", "C,B,A");

            var afterBuilder = collection.NewSnapshot("After");
            AddData(afterBuilder, "A+,B+,C+", "C+,B+,A+");

            var before = collection.GetSnapshot("Before");
            var after = collection.GetSnapshot("After");
            var diffs = SnapshotDifferenceAnalyser.ExtractDifferences(collection, before, after);

            //Act
            var result = SnapshotDifferenceSorter.SortDifferences(collection, diffs.TableDifferences.ToList());

            //Assert
            var output = new Output();
            result.Report(output);
            output.Report.Verify();
        }
    }
}