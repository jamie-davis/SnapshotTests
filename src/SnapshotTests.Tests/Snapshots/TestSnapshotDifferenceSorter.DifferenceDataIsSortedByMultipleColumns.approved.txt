TheTable                                                                                       
--------                                                                                       
              ColAsc1 ColAsc1 ColDesc1  ColDesc1 ColAsc2 ColAsc2 ColDesc2 ColDesc2 Id     Id   
Difference Id before  after   before    after    before  after   before   after    before after
---------- -- ------- ------- --------- -------- ------- ------- -------- -------- ------ -----
Updated    1  A       A+      C         C+       A       A+      C        C+       1      1    
Updated    2  A       A+      C         C+       A       A+      B        B+       2      2    
Updated    3  A       A+      C         C+       A       A+      A        A+       3      3    
Updated    4  A       A+      C         C+       B       B+      C        C+       4      4    
Updated    5  A       A+      C         C+       B       B+      B        B+       5      5    
Updated    6  A       A+      C         C+       B       B+      A        A+       6      6    
Updated    7  A       A+      C         C+       C       C+      C        C+       7      7    
Updated    8  A       A+      C         C+       C       C+      B        B+       8      8    
Updated    9  A       A+      C         C+       C       C+      A        A+       9      9    
Updated    10 A       A+      B         B+       A       A+      C        C+       10     10   
Updated    11 A       A+      B         B+       A       A+      B        B+       11     11   
Updated    12 A       A+      B         B+       A       A+      A        A+       12     12   
Updated    13 A       A+      B         B+       B       B+      C        C+       13     13   
Updated    14 A       A+      B         B+       B       B+      B        B+       14     14   
Updated    15 A       A+      B         B+       B       B+      A        A+       15     15   
Updated    16 A       A+      B         B+       C       C+      C        C+       16     16   
Updated    17 A       A+      B         B+       C       C+      B        B+       17     17   
Updated    18 A       A+      B         B+       C       C+      A        A+       18     18   
Updated    19 A       A+      A         A+       A       A+      C        C+       19     19   
Updated    20 A       A+      A         A+       A       A+      B        B+       20     20   
Updated    21 A       A+      A         A+       A       A+      A        A+       21     21   
Updated    22 A       A+      A         A+       B       B+      C        C+       22     22   
Updated    23 A       A+      A         A+       B       B+      B        B+       23     23   
Updated    24 A       A+      A         A+       B       B+      A        A+       24     24   
Updated    25 A       A+      A         A+       C       C+      C        C+       25     25   
Updated    26 A       A+      A         A+       C       C+      B        B+       26     26   
Updated    27 A       A+      A         A+       C       C+      A        A+       27     27   
Updated    28 B       B+      C         C+       A       A+      C        C+       28     28   
Updated    29 B       B+      C         C+       A       A+      B        B+       29     29   
Updated    30 B       B+      C         C+       A       A+      A        A+       30     30   
Updated    31 B       B+      C         C+       B       B+      C        C+       31     31   
Updated    32 B       B+      C         C+       B       B+      B        B+       32     32   
Updated    33 B       B+      C         C+       B       B+      A        A+       33     33   
Updated    34 B       B+      C         C+       C       C+      C        C+       34     34   
Updated    35 B       B+      C         C+       C       C+      B        B+       35     35   
Updated    36 B       B+      C         C+       C       C+      A        A+       36     36   
Updated    37 B       B+      B         B+       A       A+      C        C+       37     37   
Updated    38 B       B+      B         B+       A       A+      B        B+       38     38   
Updated    39 B       B+      B         B+       A       A+      A        A+       39     39   
Updated    40 B       B+      B         B+       B       B+      C        C+       40     40   
Updated    41 B       B+      B         B+       B       B+      B        B+       41     41   
Updated    42 B       B+      B         B+       B       B+      A        A+       42     42   
Updated    43 B       B+      B         B+       C       C+      C        C+       43     43   
Updated    44 B       B+      B         B+       C       C+      B        B+       44     44   
Updated    45 B       B+      B         B+       C       C+      A        A+       45     45   
Updated    46 B       B+      A         A+       A       A+      C        C+       46     46   
Updated    47 B       B+      A         A+       A       A+      B        B+       47     47   
Updated    48 B       B+      A         A+       A       A+      A        A+       48     48   
Updated    49 B       B+      A         A+       B       B+      C        C+       49     49   
Updated    50 B       B+      A         A+       B       B+      B        B+       50     50   
Updated    51 B       B+      A         A+       B       B+      A        A+       51     51   
Updated    52 B       B+      A         A+       C       C+      C        C+       52     52   
Updated    53 B       B+      A         A+       C       C+      B        B+       53     53   
Updated    54 B       B+      A         A+       C       C+      A        A+       54     54   
Updated    55 C       C+      C         C+       A       A+      C        C+       55     55   
Updated    56 C       C+      C         C+       A       A+      B        B+       56     56   
Updated    57 C       C+      C         C+       A       A+      A        A+       57     57   
Updated    58 C       C+      C         C+       B       B+      C        C+       58     58   
Updated    59 C       C+      C         C+       B       B+      B        B+       59     59   
Updated    60 C       C+      C         C+       B       B+      A        A+       60     60   
Updated    61 C       C+      C         C+       C       C+      C        C+       61     61   
Updated    62 C       C+      C         C+       C       C+      B        B+       62     62   
Updated    63 C       C+      C         C+       C       C+      A        A+       63     63   
Updated    64 C       C+      B         B+       A       A+      C        C+       64     64   
Updated    65 C       C+      B         B+       A       A+      B        B+       65     65   
Updated    66 C       C+      B         B+       A       A+      A        A+       66     66   
Updated    67 C       C+      B         B+       B       B+      C        C+       67     67   
Updated    68 C       C+      B         B+       B       B+      B        B+       68     68   
Updated    69 C       C+      B         B+       B       B+      A        A+       69     69   
Updated    70 C       C+      B         B+       C       C+      C        C+       70     70   
Updated    71 C       C+      B         B+       C       C+      B        B+       71     71   
Updated    72 C       C+      B         B+       C       C+      A        A+       72     72   
Updated    73 C       C+      A         A+       A       A+      C        C+       73     73   
Updated    74 C       C+      A         A+       A       A+      B        B+       74     74   
Updated    75 C       C+      A         A+       A       A+      A        A+       75     75   
Updated    76 C       C+      A         A+       B       B+      C        C+       76     76   
Updated    77 C       C+      A         A+       B       B+      B        B+       77     77   
Updated    78 C       C+      A         A+       B       B+      A        A+       78     78   
Updated    79 C       C+      A         A+       C       C+      C        C+       79     79   
Updated    80 C       C+      A         A+       C       C+      B        B+       80     80   
Updated    81 C       C+      A         A+       C       C+      A        A+       81     81   
                                                                                               